Include the new query service streaming updater (#150)
wmde/wikidata-wikibase-architecture: systems/Query/05-Building_Block_View.md

# Building Block View

## Overall System

> The Query Service is made up of a few failry distinct components.

Author: ```suggestion
The Query Service is made up of a few fairly distinct components.
```

## Web Interfaces

![](./diagrams/05-01-web-interfaces.drawio.svg)

| Building Block | Responsibility                                                                                                              |
| -------------- | --------------------------------------------------------------------------------------------------------------------------- |
| Query Frontend | Provide users with an interface to create entirly custom [SPARQL](./../../Glossary.md#sparql) queries and visualize results |
| Query Builder  | Provider users with a simple visual interface for generating lists using the query service                                  |

## Backend

![](./diagrams/05-01-backend.drawio.svg)

| Building Block | Responsibility                                                  |
| -------------- | --------------------------------------------------------------- |
| Query Proxy    | Force setting of timeouts, read only queries and basic caching  |
| Blazegraph     | Graph datagbase with a [SPARQL](./../../Glossary.md#sparql) API |
| Data Loader    | Loads data into Blazegraph from an RDF rump                     |

### Data Loader

> The Data Loader currently currently exists in 2 different flavours, which are used in difference deployments.

Reviewer: correction service: :-)

```suggestion
The Data Loader currently exists in two different flavours, which are used in different deployments.
```

> Only 1 of the choices for munging / RDF conversaion is needed per deployment.

Author: ```suggestion
Only 1 of the choices for munging / RDF conversation is needed per deployment.
```

> Only 1 of the choices for munging / RDF conversation is needed per deployment.

Reviewer: ```suggestion
Only one of the choices for munging / RDF conversation is needed per deployment.
```

> ![](./diagrams/05-02-backend-loader.drawio.svg)

Author: > This diagram is a bit misaligned. The "WikidataTurtleDumpConverter" is a bit lost in the top right corner.

Thanks, that was not meant to be there and I missed removing it in my IDE!

> About colors/content: Suddenly the Data Loader does contain persistent data as well?

This is a good question, and I'm not sure how best to handle this right now, I guess this is less data persistance and more a data transformation.
Noone cares about the fact persistence happens in this intermediary step (probably), so perhaps this is better represented in another way?
Right now though I am out of ideas!

> Shouldn't "write queries" actually be "write triples"?

For consistency this is "Write queries" everywhere right now.
Most other "write queries" for example do not only write triples, but also delete etc.
This is however a case where probably only inserts happen

| Building Block      | Responsibility                                                                                                                                             |
| ------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------- |
| loadData            | Load [RDF](./../../Glossary.md#rdf) into Blazegraph                                                                                                        |
| munger              | Convert Wikibase Repository [RDF](./../../Glossary.md#rdf) triples into different triples for Blazegraph                                                   |
| TurtleDumpConverter | Convert Wikibase Repository [RDF](./../../Glossary.md#rdf) triples into different triples for Blazegraph in a [Hadoop](./../../Glossary.md#hadoop) cluster |

## Updating

> The Updating system curently exists in 3 different flavours, which are used in different deployments.
> Only 1 of these update paths is needed in each deployment.

Reviewer: ```suggestion
The Updating system currently exists in three different flavours, which are used in different deployments.
Only one of these update paths is needed in each deployment.
```

> ![](./diagrams/05-01-updating.drawio.svg)

Author: > It would be nice if this diagram could somehow communicate that some of these flavours pull their data from Wikibase, wheras others get it pushed to them.

This is probably something better represented in the sequence diagrams, as at this level we only show data flow.
A sequence diagram for the streaming updater will probably come after this initial PR gets merged.

> And generally what the difference between the top two ones is. Besides the name, they seem almost identical

The difference here is detailed in the table just below the diagram.

> ![](./diagrams/05-01-updating.drawio.svg)

Reviewer: "Only one of these update paths is needed", but only two of the components are marked as optional and the RCPoller must always be there? If that's the case, then maybe the RCPoller deserves one more sentence in the intro paragraph above the chart. Apparently it is also the default choice for the Wikibase Docker deployment.

> ![](./diagrams/05-01-updating.drawio.svg)

Author: Good point.
Indeed, the RCPoller is more the default updater, for wikibase deployments at least.
Wikidata uses the streaming updater, and the other kafka updater is not really used at all right now afaik (probably pending removal).
Thanks for the note, I'll updater something here

| Building Block                          | Responsibility                                                                                                                           |
| --------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- |

> | [Streaming Updater](#streaming-updater) | Updating Blazegraph using a stream of [RDF](./../../Glossary.md#rdf) change events                                                       |
> | KafkaPoller                             | Updating Blazegraph using a stream of events created by [EventBus](./../../Glossary.md#eventbus) from [Kafka](./../../Glossary.md#kafka) |
> | RCPoller                                | Updating Blazegraph by polling [Recent Changes](./../../Glossary.md#recent-changes) for new changes                                      |

Reviewer: ```suggestion
| [Streaming Updater](#streaming-updater) | Update Blazegraph using a stream of [RDF](./../../Glossary.md#rdf) change events                                                       |
| KafkaPoller                             | Update Blazegraph using a stream of events created by [EventBus](./../../Glossary.md#eventbus) from [Kafka](./../../Glossary.md#kafka) |
| RCPoller                                | Update Blazegraph by polling [Recent Changes](./../../Glossary.md#recent-changes) for new changes                                      |
```

### Streaming Updater

> ![](./diagrams/05-02-streaming-updater.drawio.svg)

Author: Right now these are the names used by the Search Platform team so it's best to probably stick to these.
I'm also not sure if they do only consume and produce RDF.
Or, I believe they do, but this is actually RDF diffs as I understand.
